src/coreclr/src/jit/importer.cpp

```diff
@@ -4015,6 +4015,45 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 break;
             }
 
+            case NI_System_Type_get_IsClass:
+            case NI_System_Type_get_IsPrimitive:
+            case NI_System_Type_get_IsValueType:
+            {
+                if (impStackTop().val->IsCall())
+                {
+                    GenTreeCall* call = impStackTop().val->AsCall();
+                    if (call->gtCallMethHnd == eeFindHelper(CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE))
+                    {
+                        GenTreeIntCon* handle = call->Args().begin()->GetNode()->AsIntCon();
+                        auto hClass           = reinterpret_cast<CORINFO_CLASS_HANDLE>(handle->IconValue());
+                        typeInfo tinfo        = verMakeTypeInfo(hClass);
+                        BOOL isValueType      = tinfo.IsValueClass();
+                        if (ni == NI_System_Type_get_IsPrimitive)
+                        {
+                            if (isValueType)
+                            {
+                                retNode = gtNewIconNode(tinfo.IsPrimitiveType() ? 1 : 0);
+                            }
+                            else
+                            {
+                                retNode = gtNewIconNode(0);
+                            }
+                        }
+                        else if (ni == NI_System_Type_get_IsClass)
+                        {
+                            BOOL isInterface = info.compCompHnd->getClassAttribs(hClass) & CORINFO_FLG_INTERFACE;
```

Comment:
This won't work in shared code (`bool IsClass<T>() => typeof(T).IsClass` when T is an interface - I believe this is going to incorrectly say `true`). The code for `IsClass<T>` is going to be shared for all reference type. We don't know at compile time what the T is (at compile time the T is always the magical `System.__Canon` type for all reference types).

Reply:
As far as I understand this case is now handled in my code with `GT_RUNTIMELOOKUP` check. Perhaps, I should also check for `CORINFO_FLG_SHAREDINST` attribute...